pkg/networkservice/common/interpose/server.go

```diff
@@ -0,0 +1,166 @@
+//                        (and thus a pointer) *registry.NetworkServiceRegistryServer is a double pointer.  Meaning it
+//                        points to a place that points to a place that implements registry.NetworkServiceRegistryServer
+//                        This is done so that we can return a registry.NetworkServiceRegistryServer chain element
+//                        while maintaining the NewServer pattern for use like anything else in a chain.
+//                        The value in *server must be included in the registry.NetworkServiceRegistryServer listening
+//                        so it can capture the registrations.
+func NewServer(name string, registryServer *registry.NetworkServiceEndpointRegistryServer) networkservice.NetworkServiceServer {
+	rv := &interposeServer{
+		name: name,
+	}
+	*registryServer = interpose.NewNetworkServiceRegistryServer(&rv.endpoints)
+	return rv
+}
+
+func (l *interposeServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (result *networkservice.Connection, err error) {
+	// Check if there is no active connection, we need to replace endpoint url with forwarder url
+	conn := request.GetConnection()
+	ind := conn.GetPath().GetIndex() // It is designed to be used inside Endpoint, so current index is Endpoint already
+	connID := conn.GetId()
+
+	if len(conn.GetPath().GetPathSegments()) == 0 || ind <= 0 {
+		return nil, errors.Errorf("path segment doesn't have a client or cross connect nse identity")
+	}
+	// We need to find an Id from path to match active connection request.
+	clientConnID := l.getConnectionID(conn)
+
+	connInfoRaw, ok := l.activeConnection.Load(clientConnID)
+	if !ok {
+		if connID != clientConnID {
+			return nil, errors.Errorf("connection id should match current path segment id")
+		}
+		// We came from client, so select cross nse and go to it.
+		clientURL := clienturl.ClientURL(ctx)
+
+		// Iterate over all cross connect NSEs to check one with passed state.
+
+		l.endpoints.Range(func(key string, value *registry.NetworkServiceEndpoint) bool {
+			crossNSEURL, _ := url.Parse(value.Url)
+			crossCTX := clienturl.WithClientURL(ctx, crossNSEURL)
+
+			// Store client connection and selected cross connection URL.
+			_, _ = l.activeConnection.LoadOrStore(conn.Id, &connectionInfo{
+				endpointURL: clientURL,
+				crossNSEURL: crossNSEURL,
+			})
+			result, err = next.Server(crossCTX).Request(crossCTX, request)
+			if err != nil {
+				trace.Log(ctx).Errorf("failed to request cross NSE %v err: %v", crossNSEURL, err)
+				return true
+			}
+			// If all is ok, stop iterating.
+			return false
+		})
+		if result != nil {
+			return result, nil
+		}
+		return nil, errors.Errorf("all cross NSE failed to connect to endpoint %v connection: %v", clientURL, conn)
+	}
+
+	// Go to endpoint URL
+	connInfo := connInfoRaw.(*connectionInfo)
+	crossCTX := clienturl.WithClientURL(ctx, connInfo.endpointURL)
```

Comment:
We already have client URL set either from discover or from localbypass chain element, is there any need to set it again?

Reply:
Make seance, will fix.

Reply:
Actually It would be better to check and verify if same endpointUrl is selected on both enters.